lnschema_core/dev/sqlmodel.py

```diff
@@ -93,3 +107,100 @@ def add_relationship_keys(table: sqm.SQLModel):  # type: ignore
+            relationship_fields[field] = resolved_ann
+            pydantic_annotations[field] = (resolved_ann, None)
+        elif _is_ann_optional(ann) or _is_field_optional(model.__fields__.get(field)):
+            pydantic_annotations[field] = (resolved_ann, None)
+        else:
+            # do not validate auto-populated (server-side) created_at fields
+            if field in model.__fields__:
+                sa_column_kwargs = getattr(model.__fields__[field].field_info, "sa_column_kwargs", None)
+                if isinstance(sa_column_kwargs, dict):
+                    if "server_default" in sa_column_kwargs and field not in user_kwargs:
+                        continue
+            pydantic_annotations[field] = (resolved_ann, ...)
+
+    # pydantic does not check for the type of complex objects, only their attributes
+    for key in relationship_fields.keys() & user_kwargs.keys():
+        try:
+            check_type(key, user_kwargs[key], relationship_fields[key])
+        except Exception:
+            raise TypeError(f"Validation error for {model.__class__.__name__}: field {field} should be of type {relationship_fields[key].__name__}.")
+
+    # validate with pydantic
+    validation_model = create_model(model.__class__.__name__, **pydantic_annotations)
+    validation_model.validate(kwargs)
+    return
+
+
+def _is_ann_optional(type_ann):
+    if type_ann.__module__ == "typing":
+        try:
+            if type_ann.__origin__ == Union:
+                if type(None) in type_ann.__args__:
+                    return True
+        except Exception:
+            return False
+    return False
+
+
+def _is_field_optional(model_field):
+    if model_field is None:
+        return False
+    return not model_field.required
+
+
+def _resolve_forward_ref(ann, module):
+    # fetch target type as string
+    forward_args = []
+    parent_ref = None
+    if _is_ann_optional(ann):
+        forward_args = [ann.__args__[0].__forward_arg__]
+    else:
+        try:
+            for arg in ann.__args__:
+                if "ForwardRef" in str(arg):
+                    forward_args += [arg.__forward_arg__]
+            parent_ref = ann._name
+        except Exception:
+            forward_args = [ann.__forward_arg__]
+
+    # fetch target type from string
+    resolved_args = []
+    for arg in forward_args:
+        split_arg = arg.split(".")
+        try:
+            # if reference is in another module
+            resolved_args += [getattr(importlib.import_module(split_arg[0]), split_arg[-1])]
+        except Exception:
+            resolved_args += [getattr(importlib.import_module(module), split_arg[-1])]
+
+    # return resolved annotation
+    if len(resolved_args) == 1:
+        if parent_ref is None:
+            return resolved_args[0]
+        else:
+            return getattr(typing, parent_ref)[resolved_args[0]]
+    else:
+        return getattr(typing, parent_ref)[resolved_args[0], resolved_args[1]]
```

Comment:
Could you add a test that covers this line? It seems like something that could easily break.

Reply:
This entire conditional block is only necessary in case a field is typed with a Union-like type where none of inputs are optional (e.g. `typing.Union[str, ForwardRef("lnschema_core._core.Pipeline")]`). I created a dummy model in the test notebook with such field, but sqlmodel does not recognize non-optional Union-like fields (it breaks). Should we safely ignore this case, then?

Reply:
Meaning: Print a warning or something similar to indicate that we ignore the case. And keep the test case you created around somewhere in a note! :)

Reply:
Or maybe just as a comment in the source code. If that's easiest!

Reply:
Great! Makes things a lot simpler. Will add the warning.

Reply:
I'll add the comment because the SQLModel error is raised at class definition.

Intercepting the class definition is quite complex and would require us to front-run the instantiation of the `sqm.SQLModel` metaclass, like this (the code below works, but is useless since the SQLModel error is raised at class definition - the most we could do would be to raise an error ourselves with a custom message):

<img width="590" alt="Screen Shot 2023-01-20 at 5 06 43 PM" src="https://user-images.githubusercontent.com/33378855/213760553-07a0b0c0-1665-49b3-91c2-e90b4389e7d9.png">